docs/RFCS/20211018_index_recommendation.md

```diff
@@ -0,0 +1,207 @@
+
+### Overview
+The next step is applying the optimizer costing algorithm to determine the best set of indexes for the given workload *W*. That is, find a set of indexes *X* such that *Cost(W, X)* is minimized.
+For each statement's candidate set, determine the optimizer cost of *W* if that index subset were to be applied. Choose the statement candidate sets with the lowest *Cost(W, X)*.
+We must then check for index overlap and remove similar/identical indexes to avoid redundancy. An example of two similar indexes is having an index on `(x, y)` and then also having an index on `(x, y, z)`.
+A strategy to determine which index to remove would be running *W* again with our chosen indexes, and of the redundant indexes choose the one that has the highest worth.
+Meaning, the sum of the frequencies of the statements in which the index is used is the highest. When we re-run *W*, we should also remove any chosen indexes that are unused.
+
+At this time, potential indexes can be compared with existing indexes. If indexes we want to recommend include existing indexes, we omit those recommendations.
+In the case that no indexes remain, it means that no indexes will be recommended to the user to add.
+In a similar way, if the addition of our hypothetical indexes caused some existing indexes to become unused or rarely used, we would recommend that these indexes be deleted.
+If the index is still occasionally used, we need to ensure that removing it doesn't negatively affect overall performance. A good way to check this could be contrasting the writes to this index with the reads.
+As in, if an index is read from more than it is written to, it's better than the opposite. To fully ensure that this has not caused a regression however, we should re-run *W* with the index hypothetically dropped.
+Additionally, before we delete an index, we should ensure that no queries are regularly using hints to force that index.
+After this step, we have our optimal *X* that will be recommended to the user.
+
+In terms of the final output, we will definitely have recommended indexes. In addition, we should have a quantifiable "impact score" associate with an index recommendation that we can use to justify why the index would be beneficial to users.
+We can also include further information with this, such as what queries are affected and/or affected the most. For "drop" recommendations, we should have a similar metric.
+If an index is being frequently written to and infrequently accessed, the score would be higher than if only the latter was true.
+
+### Issues and Considerations
+One issue with this approach is that it can become a feedback loop where adding new indexes affects the existing ones, so we remove them, and then that allows for new potential indexes to be useful.
+This is a tradeoff that must be accepted - otherwise the algorithm could run infinitely. Implementation will begin with recommending potential indexes to add, followed by recommending indexes to remove.
```

Comment:
nit: do you mean "where adding new indexes affect existing **query plans**..."? Can you clarify what this tradeoff is?